tests/MigrationAPI.cpp

```diff
@@ -0,0 +1,153 @@
+        std::string first_name;
+        std::string last_name;
+        std::string email;
+    };
+
+    auto getStorage(const DbConnection& con) {
+        auto storage = make_storage(con,
+                                    make_table("users",
+                                               make_column("id", &User::id, primary_key()),
+                                               make_column("first_name", &User::first_name),
+                                               make_column("last_name", &User::last_name),
+                                               make_column("email", &User::email)));
+        return storage;
+    }
+}
+
+namespace {
+    struct OldUser {
+        int id = 0;
+        std::string name;
+    };
+}
+
+TEST_CASE("DbConnectionAPI") {
+    // insert using the original schema!
+    DbConnection con{"migration_api.sqlite"};
+    REQUIRE(con.connection.use_count() == 1);
+
+    auto oldStorage = make_storage(
+        con,
+        make_table("users", make_column("id", &OldUser::id, primary_key()), make_column("name", &OldUser::name)));
+    auto retain_count = oldStorage.get_connection_holder()->retain_count();
+    REQUIRE(retain_count == 1);
+    REQUIRE(con.connection.use_count() == 2);
+
+    // This checks whether another connection is opened - which shouldn't because all make_storages() use the DbConnection sqlite3*
+    sqlite3* pDb = nullptr;
+    oldStorage.on_open = [&pDb](sqlite3* p) {
+        pDb =
+            p;  // must never be called because on_open_internal is only called once in DbConnection's constructor (since every further call uses an open shared connection)
+    };
+    oldStorage.drop_table(oldStorage.tablename<OldUser>());
+    oldStorage.sync_schema();
+    oldStorage.pragma.user_version(1);
+    oldStorage.replace(OldUser{1, "juan dent"});
+    oldStorage.replace(OldUser{2, "klaus"});
+    oldStorage.replace(OldUser{3, "eugene"});
+    retain_count = oldStorage.get_connection_holder()->retain_count();
+    REQUIRE(retain_count == 1);
+    REQUIRE(con.connection.use_count() == 2);
+    REQUIRE(pDb == nullptr);
+
+    SECTION("register_migrations") {
+        oldStorage.register_migration(1, 2, [](const DbConnection& con) {
+            internal::connection_ref{*con.connection};
+            std::map<int, std::string> emails = {
+                {1, "fnc12@me.com"},
+                {2, "megasuperhero@gmail.com"},
+                {3, "thor@hotmail.com"},
+            };
+            auto oldStorage = version1::getStorage(con);
+            auto retain_count = oldStorage.get_connection_holder()->retain_count();
+            REQUIRE(retain_count == 1);
+            auto allOldUsers = oldStorage.get_all<version1::User>();
+            auto nextStorage = version2::getStorage(con);
+            retain_count = nextStorage.get_connection_holder()->retain_count();
+            nextStorage.sync_schema();  // old users are dropped here
+            nextStorage.pragma.user_version(2);
```

Comment:
don't set `user_version` manually cause storage can do it by itself inside `migrate_to` call